activerecord/lib/active_record/autosave_association.rb

```diff
@@ -390,11 +390,8 @@ def save_has_one_association(reflection)
           elsif autosave != false
             key = reflection.options[:primary_key] ? send(reflection.options[:primary_key]) : id
 
-            if (autosave && record.changed_for_autosave?) || new_record? || record_changed?(reflection, record, key)
-              unless reflection.through_reflection
-                record[reflection.foreign_key] = key
-              end
-
+            if (autosave && record.changed_for_autosave?) || (!reflection.through_reflection && record_changed?(reflection, record, key))
```

Comment:
I would remove the `record.new_record?` check from `record_changed?` and keep it into the if clause. This code is a hot spot, I also would suggest extracting these logical expressions into some other kind of abstraction. So :+1: @arthurnn 

I'm sorry, but the `record.new_record?` should be on the `record_changed?` method, I just missed that on the time I was working on it.